libs/execution/src/lib/types/io-types/filesystem-inmemory.ts

```diff
@@ -4,94 +4,96 @@
 
 import { IOType } from '@jvalue/jayvee-language-server';
 
-import { BinaryFile } from './binary-file';
 import { FileSystem } from './filesystem';
+import { FileSystemDirectory } from './filesystem-node-directory';
+import { FileSystemFile } from './filesystem-node-file';
+import { BinaryFile } from './filesystem-node-file-binary';
 
 export class InMemoryFileSystem implements FileSystem {
   public readonly ioType = IOType.FILE_SYSTEM;
 
+  private rootDirectory: FileSystemDirectory = new FileSystemDirectory('root');
   private static PATH_SEPARATOR = '/';
   private static CURRENT_DIR = '.';
   private static PARENT_DIR = '..';
 
-  // Hierachical file system
-  private fileSystemIndex: Map<string, BinaryFile | FileSystem> = new Map();
+  getFile(path: string): FileSystemFile<unknown> | null {
+    const processedParts = this.processPath(path);
+    let currentDir = this.rootDirectory;
 
-  private processPath(path: string): string[] {
-    const parts = path
-      .toLowerCase()
-      .split(InMemoryFileSystem.PATH_SEPARATOR)
-      .filter((p) => p !== ''); // Process paths like "folder1//folder1" to "folder1/folder2"
-    const processedParts: string[] = [];
-    for (const part of parts) {
-      if (part === InMemoryFileSystem.CURRENT_DIR) {
-        continue; // Skip current dirs in path
-      }
-      if (part === InMemoryFileSystem.PARENT_DIR) {
-        processedParts.pop(); // Go level up in folder hierarchy if ..
-      } else {
-        processedParts.push(part);
+    // Loop until we reach the last part of processedParts
+    for (let i = 0; i < processedParts.length - 1; i++) {
+      const part = processedParts[i];
+
+      // If part exists
+      if (part != null) {
+        const childNode = currentDir.getChildNode(part);
+        if (childNode instanceof FileSystemDirectory) {
+          currentDir = childNode;
+        } else {
+          return null;
+        }
       }
     }
-    return processedParts;
+    // If file exists, return
+    const child = currentDir.getChildNode(processedParts.pop() ?? '');
+    if (child !== undefined && child instanceof FileSystemFile) {
+      return child;
+    }
+    return null;
   }
 
-  getFile(filePath: string): BinaryFile | null {
-    const processedParts = this.processPath(filePath);
+  putFile(path: string, file: BinaryFile): FileSystem | null {
+    const processedParts = this.processPath(path);
```

Comment:
Similar to `findFile`, you could add a method to `FileSystemNode` that covers this functionality. Then, you could simply call `rootDirectory.putFile(path, file)` (or something similar) at this point to descent recursively and add the file at the desired location. This again would avoid the `instanceof` checks.

Reply:
My approach was to keep the FileSystemNode (composite) simple for type saftey (just add, and get methods) and move all the logic to the client in this case the `FileSystem`. 

See [wiki article](https://en.wikipedia.org/wiki/Composite_pattern#:~:text=Design%20for%20type%20safety%3A%20Child%2Drelated%20operations%20are%20defined%20only%20in%20the%20Composite%20class.%20Clients%20must%20treat%20Leaf%20and%20Composite%20objects%20differently.%20But%20type%20safety%20is%20gained%20because%20clients%20cannot%20perform%20child%2Drelated%20operations%20on%20Leaf%20objects.), which is describing design for type saftey

Reply:
I'd argue for the uniformity approach rather than the type safety approach, since this is emphasized by the composite pattern and leads to more cohesive code in my opinion. Also see [here](https://en.wikipedia.org/wiki/Composite_pattern#:~:text=The%20Composite%20design%20pattern%20emphasizes%20uniformity%20over%20type%20safety.) in the wiki article.

Reply:
Oh yes you are right, convinced me :D I changed with new commit to uniformity approach using clean recursion. I hope you like it :D